mmpy_bot/plugins/base.py

```diff
@@ -88,21 +71,185 @@ async def call_function(
             # a plugin-specific thread or process pool if we wanted.
             self.driver.threadpool.add_task(function, event, *groups)
 
-    def get_help_string(self):
-        string = f"Plugin {self.__class__.__name__} has the following functions:\n"
-        string += "----\n"
-        for functions in self.message_listeners.values():
-            for function in functions:
-                string += f"- {function.get_help_string()}"
-            string += "----\n"
-        if len(self.webhook_listeners) > 0:
-            string += "### Registered webhooks:\n"
-            for functions in self.webhook_listeners.values():
-                for function in functions:
-                    string += f"- {function.get_help_string()}"
+
+class HelpPlugin(Plugin):
+    """Provide a `help` function that lists functions provided by all plygins."""
+
+    def __init__(
+        self,
+        direct_help: bool = True,
+    ):
+        super().__init__()
+        self.direct_help: bool = direct_help
+
+    def initialize(
+        self,
+        driver: Driver,
+        manager: PluginManager,
+        settings: Settings,
+    ):
+        super().initialize(driver, manager, settings)
+
+        if self.settings.RESPOND_CHANNEL_HELP:
+            self.help = listen_to("^!help$")(self.help)
+
+    def get_help_string(self) -> str:
+        def custom_sort(rec):
+            return (
+                rec.annotations.get("category", ""),  # No categories first
+                rec.help_type,
+                rec.pattern.lstrip("^[(-"),
+            )
+
+        string = "### The following functions have been registered:\n\n"
+        string += "###### `(*)` require the use of `@botname`, "
+        string += "`(+)` can only be used in direct message\n"
+        old_category = None
+        for h in sorted(self.manager.get_help(), key=custom_sort):
```

Comment:
If we're just using `self.manager.get_help()` here, why can't we move all this logic there? Is `PluginManager.get_help` used anywhere else?

Reply:
Initially I tried implementing help on the `PluginManager` class but due to how listeners are registered the manager itself would have to behave as a `Plugin` to respond to "help". I tried going with this exception but didn't like the result so I eventually converged in `HelpPlugin`.

The separation I tried to do here is:
1. `PluginManager.get_help()` collects the help and creates `PluginHelpInfo` instances for them.
2. `HelpPlugin.get_help_string()` implement transforms the above into a string. This method can be subclassed or re-implemented in other plugins by users. 

I couldn't come up with any useful way to allow partial reuse of the `HelpPlugin.get_help_string()` code. Could maybe pull out parts of the text but it felt like too much guessing for potential use-cases.

I also think it makes more sense to have this in the plugin as an example of how to go about implementing it.